src/game/client/ingameui.cpp

```diff
@@ -63,3 +1158,597 @@ bool InGameUI::Are_Selected_Objects_Controllable()
+                if (can_select && obj->Is_Locally_Controlled()) {
+                    Set_Mouse_Cursor(CURSOR_SELECT);
+                    return;
+                }
+
+                Set_Mouse_Cursor(CURSOR_ATTACK_MOVE);
+                return;
+            case GameMessage::MSG_ADD_WAYPOINT_HINT:
+                Set_Mouse_Cursor(CURSOR_WAYPOINT);
+                return;
+            case GameMessage::MSG_HIJACK_HINT:
+            case GameMessage::MSG_SABOTAGE_HINT:
+            case GameMessage::MSG_CONVERT_TO_CARBOMB_HINT:
+                Set_Mouse_Cursor(CURSOR_ENTER_AGRESSIVE);
+                return;
+            case GameMessage::MSG_CAPTUREBUILDING_HINT:
+                Set_Mouse_Cursor(CURSOR_CAPTURE_BUILDING);
+                return;
+            case GameMessage::MSG_DEFECTOR_HINT:
+                Set_Mouse_Cursor(CURSOR_DEFECTOR);
+                return;
+            case GameMessage::MSG_SET_RALLY_POINT_HINT:
+                if (can_select) {
+                    Set_Mouse_Cursor(CURSOR_SELECT);
+                    return;
+                }
+
+                Set_Mouse_Cursor(CURSOR_SET_RALLY_POINT);
+                return;
+            case GameMessage::MSG_UNK2:
+                Set_Mouse_Cursor(CURSOR_PARTICLE_UPLINK_CANNON);
+                return;
+            case GameMessage::MSG_DO_SALVAGE_HINT:
+                Set_Mouse_Cursor(CURSOR_MOVE);
+                break;
+            case GameMessage::MSG_DO_ATTACK_OBJECT_AFTER_MOVING_HINT:
+                Set_Mouse_Cursor(CURSOR_OUT_RANGE);
+                return;
+            case GameMessage::MSG_HACK_HINT:
+                Set_Mouse_Cursor(CURSOR_HACK);
+                return;
+            default:
+                return;
+        }
+    }
+}
+
+void InGameUI::Set_GUI_Command(CommandButton const *command)
+{
+    if (g_theRecorder->Get_Mode() == RECORDERMODETYPE_PLAYBACK) {
+        return;
+    }
+
+    if (command != nullptr) {
+        if ((command->Get_Options()
+                & (COMMAND_OPTION_CONTEXTMODE_COMMAND | COMMAND_OPTION_NEED_TARGET_POS
+                    | COMMAND_OPTION_NEED_TARGET_ALLY_OBJECT | COMMAND_OPTION_NEED_TARGET_NEUTRAL_OBJECT
+                    | COMMAND_OPTION_NEED_TARGET_ENEMY_OBJECT))
+            == 0) {
+            captainslog_dbgassert(
+                "Set_GUI_Command: Command '%s' does not need additional user interaction", command->Get_Name().Str());
+            m_pendingGUICommand = nullptr;
+            m_mouseMode = MOUSEMODE_DEFAULT;
+            return;
+        }
+
+        m_mouseMode = MOUSEMODE_GUI_COMMAND;
+    } else {
+        m_mouseMode = MOUSEMODE_DEFAULT;
+    }
+
+    m_pendingGUICommand = command;
+
+    if (command != nullptr
+        && (command->Get_Options()
+               & (COMMAND_OPTION_CONTEXTMODE_COMMAND | COMMAND_OPTION_NEED_TARGET_POS
+                   | COMMAND_OPTION_NEED_TARGET_ALLY_OBJECT | COMMAND_OPTION_NEED_TARGET_NEUTRAL_OBJECT

```

Comment:
Or combine last 3 to COMMAND_OPTION_NEED_TARGET_OBJECT

Reply:
fixed.